### PR TITLE
[Passes] Run SimpleLoopUnswitch after introducing invariant branches.

### LINKED ISSUE #85551: Indexing bounds checks are not hoisted out of a loop
I played around with some variants on using std::span bounds checks, and it seems Clang is not yet very good at hoisting checks out of loops.

Consider a (slightly silly) function like this:
```
int sum_prefix(std::span<const int> s, size_t n) {
    int ret = 0;
    for (size_t i = 0; i < n; i++) {
        ret += s[i];
    }
    return ret;
}
```

With bounds checks on, `s[i]` will check that `i < s.size()` at every loop iteration. But ideally we'd do a single check that `n <= s.size()` once, before the loop. That then allows Clang to vectorize the loop and do other things.

Here's a godbolt with some more illustrative variations:
https://godbolt.org/z/zYbfzco4e

### PR DESCRIPTION
**Pull request by fhahn**
IndVars may be able to replace a loop dependent condition with a loop invariant one, but loop-unswitch runs before IndVars, so the invariant check remains in the loop.

For an example, consider a read-only loop with a bounds check: https://godbolt.org/z/8cdj4qhbG

This patch uses a approach similar to the way extra cleanup passes are run on demand after vectorization (added in acea6e9cfa4c4a0e8678c7).

It introduces a new ShouldRunExtraSimpleLoopUnswitch analysis marker, which IndVars can use to indicate that extra unswitching is beneficial.

ExtraSimpleLoopUnswitchPassManager uses this analysis to determine whether to run its passes on a loop.

Compile-time impact (geomean) ranges from +0.0% to 0.02% https://llvm-compile-time-tracker.com/compare.php?from=138c0beb109ffe47f75a0fe8c4dc2cdabe8a6532&to=19e6e99eeb280d426907ea73a21b139ba7225627&stat=instructions%3Au

Compile-time impact (geomean) of unconditionally running SimpleLoopUnswitch ranges from +0.05% - +0.16%
https://llvm-compile-time-tracker.com/compare.php?from=138c0beb109ffe47f75a0fe8c4dc2cdabe8a6532&to=2930dfd5accdce2e6f8d5146ae4d626add2065a2&stat=instructions:u

Unconditionally running SimpleLoopUnswitch seems to indicate that there are multiple other scenarios where we fail to run unswitching when opportunities remain.

Fixes https://github.com/llvm/llvm-project/issues/85551.

### CODE REVIEW

File: llvm/include/llvm/Transforms/Utils/SimplifyIndVar.h [hunk 1]
```diff
@@ -47,10 +49,12 @@ class IVVisitor {
 
 /// simplifyUsersOfIV - Simplify instructions that use this induction variable
 /// by using ScalarEvolution to analyze the IV's recurrence.
```
Comment (alinas) [inline llvm/include/llvm/Transforms/Utils/SimplifyIndVar.h:51]: Add comment on what the new return value represents.

File: llvm/lib/Transforms/Utils/SimplifyIndVar.cpp [hunk 2]
```diff
@@ -981,14 +984,15 @@ void IVVisitor::anchor() { }
 
 /// Simplify instructions that use this induction variable
 /// by using ScalarEvolution to analyze the IV's recurrence.
-bool simplifyUsersOfIV(PHINode *CurrIV, ScalarEvolution *SE, DominatorTree *DT,
-                       LoopInfo *LI, const TargetTransformInfo *TTI,
-                       SmallVectorImpl<WeakTrackingVH> &Dead,
-                       SCEVExpander &Rewriter, IVVisitor *V) {
```
Comment (alinas) [inline llvm/lib/Transforms/Utils/SimplifyIndVar.cpp:987]: Same here, add a comment on what the pair return stands for.

**Review by alinas — COMMENTED** [review]
This brings back the discussion that we really should have better infrastructure in the NPM to specify the conditional running of passes. The `extra-vectorizer-passes` change was supposed to be a one time approach.

Do you have a measure of how this affects execution time for some workloads (e.g. spec)? Could you also test this on https://github.com/google/fleetbench?

**Comment by fhahn** [comment]
> This brings back the discussion that we really should have better infrastructure in the NPM to specify the conditional running of passes. The `extra-vectorizer-passes` change was supposed to be a one time approach.
> 

Sure, I am not really tied to the current approach and would be happy to explore alternative options; the current implementation was just the easiest way to get this done I was aware of.

> Do you have a measure of how this affects execution time for some workloads (e.g. spec)? Could you also test this on https://github.com/google/fleetbench?

Not yet, but I am planning to do this. I think the bounds check example highlights a good missed indvars -> SLU interaction

**Comment by fhahn** [comment]
@alinas @aeubanks any preferences/suggestions on how to better handle this in general?

**Comment by fhahn** [comment]
ping :)

**Comment by fhahn** [comment]
> So, is it an option to add IndVarSimplify to LPM1? (I don't think it makes any changes that would invalidate MSSA, so that should be fairly straightforward to add) Does it address the use case you are looking at? And following up with the same question as for the current approach: what's the compile time vs execution time impact then? i.e. is it worth adding it?

I finally had time to try this out in https://github.com/llvm/llvm-project/commit/aff602c6f3d9b66d7539d72e9f78adbf49118890, apologies for the long delay.

The impact of moving `indVarSimplify` to LPM1 is much bigger, both in terms of compile-time and binary changes: https://llvm-compile-time-tracker.com/compare.php?from=1d06f41b72e429a5b3ba318ff639b8b997e21ff8&to=aff602c6f3d9b66d7539d72e9f78adbf49118890&stat=instructions:u. I think a number of optimizations went into`IndVarSimplify` to make it work very well at the specific position. There may be some interesting insights from some of the bigger changes caused by moving it, but it seems moving it to LPM1 would be quite a heavy hammer. 

Exposing the unswitching logic directly and use that on demand in `IndVarSimplify` would likely be the most efficient approach. It also compared using a dedicated pass vs directly invoking functionality on demand in #83860 and directly invocation was noticeably faster. Should we go down this route?

**Comment by nikic** [comment]
In the interest of not making perfect the enemy of good, I'd be happy to go forward with your current approach.

Could you please resolve the comments regarding extra comments and check whether this patch fixes https://github.com/llvm/llvm-project/issues/85551 (and if so add another test)?

File: llvm/lib/Transforms/Utils/SimplifyIndVar.cpp (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Transforms/Utils/SimplifyIndVar.cpp:999]: Added, thanks!

File: llvm/include/llvm/Transforms/Utils/SimplifyIndVar.h (hunk 1 above)
Comment (fhahn) [reply llvm/include/llvm/Transforms/Utils/SimplifyIndVar.h]: Added, thanks!

**Review by fhahn — COMMENTED** [review]
> Could you please resolve the comments regarding extra comments and check whether this patch fixes https://github.com/llvm/llvm-project/issues/85551 (and if so add another test)?

Thanks, I added the comment and check that it fixes  https://github.com/llvm/llvm-project/issues/85551. Added a test case in 94ed57dab64ccb248a342a91957f390209c5c7ce.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**